File: backend/decks/stats.py
from datetime import timedelta

from django.db.models import Count, Q, Sum
from django.db.models.functions import TruncDate
from django.utils import timezone

from .models import Card, Deck, Review
from .scheduling import PASS_THRESHOLD

MATURE_INTERVAL_DAYS = 21
RECENT_WINDOW_DAYS = 30


def _ratio(passed, total):
    return round(passed / total, 4) if total else None


def _daily_counts(queryset, field):
    rows = (
        queryset.annotate(day=TruncDate(field))
        .values("day")
        .annotate(count=Count("id"))
    )
    return {row["day"]: row["count"] for row in rows}


def _dense_series(counts, start, days):
    return [
        {
            "date": (start + timedelta(days=offset)).isoformat(),
            "count": counts.get(start + timedelta(days=offset), 0),
        }
        for offset in range(days)
    ]


def _streaks(series, today):
    active = {row["date"] for row in series if row["count"] > 0}

    cursor = today
    if cursor.isoformat() not in active:
        cursor -= timedelta(days=1)

    current = 0
    while cursor.isoformat() in active:
        current += 1
        cursor -= timedelta(days=1)

    longest = 0
    run = 0
    for row in series:
        run = run + 1 if row["count"] > 0 else 0
        longest = max(longest, run)

    return {"current": current, "longest": longest}


def _weekly_insights(reviews, heatmap, today, weeks=8):
    """Return aligned retention and activity buckets for the recent weeks."""
    start = today - timedelta(days=(weeks * 7) - 1)
    review_rows = {
        row["day"]: row
        for row in (
            reviews.filter(reviewed_at__date__gte=start)
            .annotate(day=TruncDate("reviewed_at"))
            .values("day")
            .annotate(
                count=Count("id"),
                passed=Count("id", filter=Q(grade__gte=PASS_THRESHOLD)),
            )
        )
    }
    activity = {row["date"]: row["count"] for row in heatmap}
    retention_trend = []
    streak_history = []
    for week in range(weeks):
        week_start = start + timedelta(days=week * 7)
        dates = [week_start + timedelta(days=offset) for offset in range(7)]
        count = sum(review_rows.get(day, {}).get("count", 0) for day in dates)
        passed = sum(review_rows.get(day, {}).get("passed", 0) for day in dates)
        retention_trend.append({
            "date": week_start.isoformat(),
            "reviews": count,
            "retention": _ratio(passed, count),
        })
        streak_history.append({
            "date": week_start.isoformat(),
            "active_days": sum(1 for day in dates if activity.get(day.isoformat(), 0) > 0),
            "reviews": sum(activity.get(day.isoformat(), 0) for day in dates),
        })
    return retention_trend, streak_history


def build_stats(user, history_days=365, horizon_days=30):
    now = timezone.now()
    today = timezone.localdate()

    reviews = Review.objects.filter(card__deck__owner=user)
    cards = Card.objects.filter(deck__owner=user)

    all_time = reviews.aggregate(
        total=Count("id"),
        passed=Count("id", filter=Q(grade__gte=PASS_THRESHOLD)),
    )
    recent = reviews.filter(reviewed_at__gte=now - timedelta(days=RECENT_WINDOW_DAYS)).aggregate(
        total=Count("id"),
        passed=Count("id", filter=Q(grade__gte=PASS_THRESHOLD)),
    )
    card_totals = cards.aggregate(
        total=Count("id"),
        mature=Count("id", filter=Q(interval_days__gte=MATURE_INTERVAL_DAYS)),
        lapses=Sum("lapses"),
    )

    history_start = today - timedelta(days=history_days - 1)
    heatmap = _dense_series(
        _daily_counts(reviews.filter(reviewed_at__date__gte=history_start), "reviewed_at"),
        history_start,
        history_days,
    )

    forecast = _dense_series(
        _daily_counts(
            cards.filter(due_at__gt=now, due_at__date__lte=today + timedelta(days=horizon_days - 1)),
            "due_at",
        ),
        today,
        horizon_days,
    )

    backlog = cards.filter(Q(due_at__isnull=True) | Q(due_at__lte=now)).count()

    retention_trend, streak_history = _weekly_insights(reviews, heatmap, today)
    deck_rows = reviews.values(
        "card__deck_id", "card__deck__title", "card__deck__emoji"
    ).annotate(
        reviews=Count("id"),
        passed=Count("id", filter=Q(grade__gte=PASS_THRESHOLD)),
    )
    weakest_decks = sorted([
        {
            "id": row["card__deck_id"],
            "title": row["card__deck__title"],
            "emoji": row["card__deck__emoji"],
            "reviews": row["reviews"],
            "retention": _ratio(row["passed"], row["reviews"]),
        }
        for row in deck_rows
    ], key=lambda row: (row["retention"], -row["reviews"]))[:5]

    difficult_cards = list(
        cards.annotate(
            review_count=Count("reviews"),
            failed_reviews=Count("reviews", filter=Q(reviews__grade__lt=PASS_THRESHOLD)),
        )
        .filter(review_count__gt=0)
        .order_by("-lapses", "-failed_reviews", "mastery", "id")
        .values("id", "deck_id", "deck__title", "front", "lapses", "failed_reviews", "review_count")[:8]
    )

    return {
        "totals": {
            "reviews": all_time["total"],
            "cards": card_totals["total"],
            "mature_cards": card_totals["mature"],
            "lapses": card_totals["lapses"] or 0,
            "decks": Deck.objects.filter(owner=user).count(),
        },
        "retention": {
            "all_time": _ratio(all_time["passed"], all_time["total"]),
            "recent": _ratio(recent["passed"], recent["total"]),
            "recent_reviews": recent["total"],
            "window_days": RECENT_WINDOW_DAYS,
        },
        "streak": _streaks(heatmap, today),
        "backlog": backlog,
        "heatmap": heatmap,
        "forecast": forecast,
        "retention_trend": retention_trend,
        "streak_history": streak_history,
        "weakest_decks": weakest_decks,
        "difficult_cards": difficult_cards,
    }

File: frontend/src/services/stats.js
import { api, USE_MOCK_API, request } from './api'
import { isDue } from '../lib/sm2'

const MATURE_INTERVAL_DAYS = 21

function isoDay(date) {
  const year = date.getFullYear()
  const month = String(date.getMonth() + 1).padStart(2, '0')
  const day = String(date.getDate()).padStart(2, '0')
  return `${year}-${month}-${day}`
}

function denseDays(start, count, counts) {
  return Array.from({ length: count }, (_, offset) => {
    const date = new Date(start)
    date.setDate(date.getDate() + offset)
    const key = isoDay(date)
    return { date: key, count: counts[key] || 0 }
  })
}

// Demo mode has no server-side review log. Totals, backlog and forecast are
// computed from real local data; the heatmap is synthetic so the page has
// something to show.
async function mockStats({ days = 365, horizon = 30 } = {}) {
  const decks = await api.listDecks()
  const cards = decks.flatMap((deck) => deck.cards)
  const now = Date.now()

  const history = {}
  let seed = 7
  for (let offset = 0; offset < days; offset += 1) {
    const date = new Date()
    date.setDate(date.getDate() - offset)
    seed = (seed * 1103515245 + 12345) % 2147483648
    const roll = seed % 100
    if (roll > 38) history[isoDay(date)] = 3 + (seed % 22)
  }

  const start = new Date()
  start.setDate(start.getDate() - (days - 1))
  const heatmap = denseDays(start, days, history)

  const upcoming = {}
  cards.forEach((card) => {
    if (!card.dueAt) return
    const due = new Date(card.dueAt)
    if (due.getTime() <= now) return
    const key = isoDay(due)
    upcoming[key] = (upcoming[key] || 0) + 1
  })

  const reviews = heatmap.reduce((sum, day) => sum + day.count, 0)
  const recentWeeks = Array.from({ length: 8 }, (_, week) => {
    const slice = heatmap.slice(-(8 - week) * 7, -(7 - week) * 7 || undefined)
    const weeklyReviews = slice.reduce((sum, day) => sum + day.count, 0)
    return {
      date: slice[0]?.date || isoDay(new Date()),
      reviews: weeklyReviews,
      retention: weeklyReviews ? 0.82 + (week % 4) * 0.03 : null,
      active_days: slice.filter((day) => day.count > 0).length,
    }
  })
  const weakestDecks = decks.map((deck) => ({
    id: deck.id,
    title: deck.title,
    emoji: deck.emoji,
    reviews: deck.cards.reduce((sum, card) => sum + Math.max(1, card.repetitions || 0), 0),
    retention: deck.cards.length
      ? deck.cards.reduce((sum, card) => sum + card.mastery, 0) / deck.cards.length
      : null,
  })).filter((deck) => deck.retention !== null).sort((left, right) => left.retention - right.retention).slice(0, 5)
  const difficultCards = cards.map((card) => ({
    id: card.id,
    deck_id: decks.find((deck) => deck.cards.some((item) => item.id === card.id))?.id,
    deck__title: decks.find((deck) => deck.cards.some((item) => item.id === card.id))?.title,
    front: card.front,
    lapses: card.lapses || 0,
    failed_reviews: card.lapses || 0,
    review_count: card.repetitions || 0,
    mastery: card.mastery,
  })).sort((left, right) => right.lapses - left.lapses || left.mastery - right.mastery).slice(0, 8)

  return {
    totals: {
      reviews,
      cards: cards.length,
      mature_cards: cards.filter((card) => card.intervalDays >= MATURE_INTERVAL_DAYS).length,
      lapses: cards.reduce((sum, card) => sum + (card.lapses || 0), 0),
      decks: decks.length,
    },
    retention: {
      all_time: 0.87,
      recent: 0.91,
      recent_reviews: heatmap.slice(-30).reduce((sum, day) => sum + day.count, 0),
      window_days: 30,
    },
    streak: streaksFrom(heatmap),
    backlog: cards.filter((card) => isDue(card, now)).length,
    heatmap,
    forecast: denseDays(new Date(), horizon, upcoming),
    retention_trend: recentWeeks.map(({ date, reviews: count, retention }) => ({ date, reviews: count, retention })),
    streak_history: recentWeeks.map(({ date, reviews: count, active_days }) => ({ date, reviews: count, active_days })),
    weakest_decks: weakestDecks,
    difficult_cards: difficultCards,
  }
}

function streaksFrom(series) {
  const active = new Set(series.filter((day) => day.count > 0).map((day) => day.date))
  const cursor = new Date()
  if (!active.has(isoDay(cursor))) cursor.setDate(cursor.getDate() - 1)

  let current = 0
  while (active.has(isoDay(cursor))) {
    current += 1
    cursor.setDate(cursor.getDate() - 1)
  }

  let longest = 0
  let run = 0
  series.forEach((day) => {
    run = day.count > 0 ? run + 1 : 0
    longest = Math.max(longest, run)
  })

  return { current, longest }
}

function validateStats(payload) {
  const valid = payload
    && payload.totals
    && payload.retention
    && payload.streak
    && Array.isArray(payload.heatmap)
    && Array.isArray(payload.forecast)
    && Array.isArray(payload.retention_trend)
    && Array.isArray(payload.streak_history)
    && Array.isArray(payload.weakest_decks)
    && Array.isArray(payload.difficult_cards)
    && typeof payload.backlog === 'number'
  if (!valid) throw new Error('The server returned invalid progress data.')
  return payload
}

export async function getStats({ days = 365, horizon = 30 } = {}) {
  const payload = USE_MOCK_API
    ? await mockStats({ days, horizon })
    : await request(`/api/stats/?days=${days}&horizon=${horizon}`)
  return validateStats(payload)
}
